### PR TITLE
Cache the compiled module in sync MODULARIZE

### PR DESCRIPTION
**Pull request by kripken**
In MODULARIZE mode we provide a factory function from which
many instances can be created. On the Web there is automatic
caching of compiled modules, but on node there isn't. At least
when wasm compilation is sync, it's easy to cache the compiled
module, by saving it in the outermost scope (outside of the scope
of the factory function itself, a new one of which is created for
each instance).

When compilation is async, caching is trickier as there can be
races. But also that case likely matters less since again, on
the web, where we must be async, we have browser
code caching anyhow.

(preamble.js change is almost all whitespace)

### CODE REVIEW

**Comment by kripken** [comment]
cc @lourd @curiousdannii

File: emcc.py [hunk 1]
```diff
@@ -3421,18 +3427,18 @@ def modularize():
...
+  %(lines)s
 })();
-''' % {
-      'EXPORT_NAME': shared.Settings.EXPORT_NAME,
-      'script_url': script_url,
-      'script_url_node': script_url_node,
-      'src': src
-    }
+''' % {'EXPORT_NAME': shared.Settings.EXPORT_NAME,
+       'lines': '\n  '.join(lines)}
```
Comment (sbc100) [inline emcc.py:3441]: I wonder if this function could be moved to a library JS file where the `if CACHE_MODULARIZE_MODULE` pattern might be more readable?

Comment (lourd) [reply emcc.py:3441]: +1, this is hard to follow. Or maybe just adding the `var cachedModule;` statement using string substitution instead of introducing the array and having to follow the appending and joining.

**Comment by lourd** [comment]
Thanks for the CC! Looks ok to me.

Is it in Node's plans or at least on their radar to add caching for WASM modules?

**Comment by curiousdannii** [comment]
I don't know the full details of this PR, but I wouldn't have thought race conditions would be a problem. `WebAssembly.compile()` returns a promise, so all calls to the module factory can await that.

**Comment by curiousdannii** [comment]
Checking I understand: the problem with MODULARIZE is that it wraps the entire program, so there's no static closure in which to keep a cached wasm module?

(This makes me frustrated with MODULARIZE again lol. I'm becoming more convinced it's an anti-pattern.)

Rather than caching to the file system, why not put it on `global`? Here's a rough idea:

1. emit a UUID into each build so each module can have a way to refer to itself
2. Use `Object.defineProperty` to add a `wasmUUID` property onto `global` or `window`, with the value being the promise returned from `WebAssembly.compile`
3. In a subsequent call to the factory, check for the existence of `global.wasmUUID`, and if so await it rather than doing step 2

Are you sure all browsers cache modules automatically? You could use the same code in browsers if they don't.

**Comment by curiousdannii** [comment]
Ah, I saw the code for `NODE_CODE_CACHING`, but now I see that's not really being changed, it was doing that before.

What do you think about making `cachedModule` the promise returned from `WebAssembly.compile` in async mode?

**Comment by kripken** [comment]
> What do you think about making cachedModule the promise returned from WebAssembly.compile in async mode?

I don't understand - what does "make cachedModule the promise" mean here? (replace the promise with cachedModule?)

**Comment by curiousdannii** [comment]
Well in async mode you said there are races, I assume you're thinking of cases where you call the factory multiple times before the first invocation has finished compiling the wasm module?

That's easy to handle by storing the promise returned from `WebAssembly.compile` in `cachedModule` instead of the compiled module.

**Comment by kripken** [comment]
Interesting idea to cache the promise - we can consider that as a followup perhaps?

File: emcc.py (hunk 1 above)
Comment (lourd) [reply emcc.py:3441]: This is absolutely bikeshedding, but I do still think it's clearer as string substitutions.

**Comment by lourd** [comment]
Caching the promise means that follow-on calls to the factory will get the same module instance. Not sure that's the desired behavior here? I also think that would also be simple enough for users to handle themselves.

**Comment by curiousdannii** [comment]
Is it a problem to use the same wasm module? When you call `WebAssembly.instantiate` you can pass in the memory in the import object. I assume this is the intended purpose, to let you compile a module once, and then instantiate it multiple times.

File: emcc.py (hunk 1 above)
Comment (kripken) [reply emcc.py:3441]: Fair enough. Maybe the non-bikeshedding part can decide it, then: is there a clean way to avoid extra empty lines with string substitutions?

Comment (lourd) [reply emcc.py:3441]: Can't think of anything aside from putting the newline character in the substituted string

```python
  src = '''
var %(EXPORT_NAME)s = (function() {	var %(EXPORT_NAME)s = (function() {
  var _scriptDir = %(script_url)s;	  %(lines)s
  %(script_url_node)s%(cached_module_var)s
  return (%(src)s);	
})();	})();
''' % {
      'EXPORT_NAME': shared.Settings.EXPORT_NAME,
      'script_url': script_url,	
      'script_url_node': script_url_node,	
      'cached_module_var': '\nvar cachedModule;' if shared.Settings.CACHE_MODULARIZE_MODULE else '',
      'src': src	
    }
```

**Comment by kripken** [comment]
I think I'll give up on this change for now - seems controversial in the details, and I don't have time to focus on it.
